src/v/cloud_topics/reader/placeholder_extent_reader.cc

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_topics/reader/placeholder_extent_reader.h"
+
+#include "cloud_io/basic_cache_service_api.h"
+#include "cloud_io/io_result.h"
+#include "cloud_io/remote.h"
+#include "cloud_topics/dl_placeholder.h"
+#include "cloud_topics/errc.h"
+#include "cloud_topics/logger.h"
+#include "cloud_topics/reader/placeholder_extent.h"
+#include "model/fundamental.h"
+#include "model/record_batch_reader.h"
+#include "model/record_batch_types.h"
+
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/io_priority_class.hh>
+#include <seastar/core/iostream.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/coroutine/as_future.hh>
+
+#include <chrono>
+#include <exception>
+
+using namespace std::chrono_literals;
+
+namespace experimental::cloud_topics {
+
+/// Convert record batch to dl_placeholder value and return an empty
+/// extent which later has to be hydrated.
+size_t get_extent_size(const model::record_batch& placeholder) {
+    /// FIXME: This code makes unnecessary copy
+    vassert(
+      placeholder.header().type == model::record_batch_type::dl_placeholder,
+      "Unsupported batch type {}",
+      placeholder.header());
+    iobuf payload = placeholder.data().copy();
+    iobuf_parser parser(std::move(payload));
+    auto record = model::parse_one_record_from_buffer(parser);
+    iobuf value = std::move(record).release_value();
+    auto p = serde::from_iobuf<dl_placeholder>(std::move(value));
+    return p.size_bytes;
+}
+
+inline ss::future<result<ss::circular_buffer<placeholder_extent>>>
+materialize_sorted_run(
+  ss::circular_buffer<model::record_batch> placeholders,
+  cloud_storage_clients::bucket_name bucket,
+  cloud_io::remote_api<>* api,
+  cloud_io::basic_cache_service_api<>* cache,
+  retry_chain_node* rtc) {
+    absl::node_hash_map<uuid_t, ss::lw_shared_ptr<hydrated_L0_object>> hydrated;
+    ss::circular_buffer<placeholder_extent> extents;
+    for (auto&& p : placeholders) {
```

Comment:
why is this auto&&?

Reply:
it's moved in the loop body

Reply:
@Lazin 

> it's moved in the loop body

but `auto&&` isn't an r-value reference, right? it's a forwarding reference, but there isn't any type deduction happening here?